### ISSUE TITLE
One to Many and Many to Many Relationships When There are Duplicate Tables Needed

### ISSUE THREAD

**Issue by dukesnuz** [susanBuck/dwa15-fall2017#240]
Hi
I Just wanted to post this and get some feedback and your thoughts on this topic.

### Scenario
There are the following tables:
Table A, Table B, Table C

Each table needs an addresses table and a contacts table( each Table is identical rows names ).
The contacts table needs a  many to many relationship to campaigns table( each campaigns table has identical rows names ).

### My Question
Should each table have their own addresses table and their own contacts table? Table A, Table B and Table C would have a one to many relationship to their specific addresses table and the contacts table.
Each contacts table has a many to many relationship to a campaigns table using a pivot table.
Thus having 15 tables

or

Should there be a total of one addresses table and one contacts table and one campaigns table?  Table A, Table B and Table C would have a one to many relationship to the addresses table and the contacts table. The contact table would have a many to many relationship to the campaigns table using a pivot table.
Thus having a total of 7 tables.

Thank you

**Comment by ysrinu** [comment]
Hi @dukesnuz , table cannot have another table. I guess what you meant here is a column in TableA points to Table B and another column to Table C.
I don't see how you can have one-to-many for Table A to Address table. You can have many-to-one though. For example, if Table A has student records, (primary on student id), the address column (FK) can point to only one address record in the Address table.

If you can indicate what kind of data you are storing in these tables, then it will be easy to help.
Again, the database design depends on the data type and business requirements.

**Comment by dukesnuz** [comment]
@ysrinu Thank you for the reply

This is what I was trying to convey above.
There are 3 tables: Shippers, Carriers, Dispatchers

Each of the above can have more than one address and more than one people contacts.
I was thinking of a one to many relationship. So my question is would it be best to have a shipper_address table, carrier_address table and a dispatchers_address table and 
shipper_contact table, carrier_contact table, and a dispatcher_contact table ?
or
Would it be best to have  only one address table and only one contacts table and then have 
Shippers, Carriers, Dispatchers each have a one to many relationship to the one and only address and the one and only contacts table?

2nd part  each contact has  a many to many relationship to the campaigns table using a pivot table
Thus  Would it be best to have a shipper_campaigns table, carrier_campaign_table and a dispatcher_address table and each has its own pivot table
or
Only have one campaign table and one pivot table for shipper, dispatcher and carrier tables.

Here is what I have thus far.

![administrator mysql diagram](https://user-images.githubusercontent.com/3799157/35049734-fb3636b2-fb6e-11e7-8c0a-34dff9b448a6.png)

**Comment by ysrinu** [comment]
Will address used by Shipper has a different context than the address used by Carrier?
Is the address used by Shipper about the Shipper's address (and not shipping/customer address)?
Is address used by Carrier about it's own address?

If all the 3 depts are not using customer's address table, then should have different address tables, because their context is different.

If a Shipper has multiple addresses (assuming these address for Shipper's locations and not relevant for Carrier and Dispatcher). Then you should have ShipAdd table to store just these addresses.
You should also have Shipper-ShipAdd table to indicate that one shipper (shipper_id) can relate to multiple addresses (shipadd_id)

In case you want to define a default address (since many addresses for a given shipper_id is possible) you can set a flag for default relation in the Shipper-ShipAdd. This is not very useful storage usage, since the rest of the records for the same shipper would have flag off. An alternative way is to add a column in the Shipper table that will indicate the default shipper address record. This alternative can increase complexity when address table is updated, because you may have to then update both the Shipper and Shipper-ShipAdd tables too.

Use the same criteria for Carrier and Dispatcher.

Ideally, I would avoid the name 'address' since Shipper's address can easily be misread/misinterpreted for Shipping address. I would say Shipping Location instead. But why would you have multiple locations for a shipper? If these are branches or freight service locations for the shipper, then it makes sense.

**Comment by SridharSubramaniam** [comment]
Addresses are unique by nature - whether it is used for personal
purpose(contact) or business. I would recommend you keep one address table
and associate it to to appropriate entity either via one to one or pivot
table depends on use case.   Keeping all addresses in one place would help
it to keep the addresses clean and have centralize business flows in
picking/validating addresses. (e.g Google Maps lookup, GeoLocation Lookup
etc)

**Comment by ysrinu** [comment]
From data perspective addresses can belong to one table.
Here, however, it makes sense to separate them into different tables based on business. This makes them more secure too. A shipper's portal and middle-tier rules can have restriction so they cannot access Carrier or Dispatcher's table. This is much easy to manage than to filter at records level in one all-in-one table.
For example, you shouldn't store customer's and manufacturer's address in one table, even if the data type is the same, because it makes it easy to set security at table level.
With a single table, someone can change some query parameters or cross-site forgery and introduce values such the query now runs for address belonging to different department.
Separate tables, with a service account having specific privilege to a specific table or based on the originating user id or requestor session will ensure that it cannot access table you are not authorized to.

Independent tables also makes it easy to query, for a shipper's transaction it's only read/update/delete it's table which would be must faster than doing the same operation on a table that has unrelated data too (Carrier and Dispatcher). Even though a transaction can be for a single record, it's takes computing resources to build/maintain indexes and search through index too.
This scales better too.

Also, think about future data model changes. A shipper address might need new columns that may not be applicable to Carrier and Dispatcher. If we used a single table for all, then these new columns will be always be empty for Carrier/Dispatcher and is waste of database space and computing resources.

Separate tables also makes it easy to import/export data, because you don't have to worry about filtering out based on dept type. And if there's a corruption, the risk is mitigated to that particular dept data alone

Going further, when you schedule backups or database statistics, only one table at a time can have a mild performance hit during that window, instead of all depts (if single table) transactions experiencing a performance issues.

A single address table sounds good, until the tables grow rapidly, the data properties needed for each department now varies so much that a single table structure doesn't scale well for all the 3 departments.

**Comment by SridharSubramaniam** [comment]
Party Model depicts it very well that is used by large enterprise products.
A party is a party that you do business with - shipper or organization or
consumer. An address is an address of party. You can build specific
functionality on concrete entities related to each.

https://docs.oracle.com/cd/E18283_01/doc.112/e15886/logical_cdm.htm#CACJIJEI

**Comment by ysrinu** [comment]
@SridharSubramaniam I think you are referring to object-relational model.
(Entity Framework is the term used in Microsoft .NET Framework)
At the database level, it's still a relational model alone (assuming you are using RDBMS system and not document databases) and not object-relation.

In your application, you can define classes and use inheritance and table mappings to retrieve the right data and organize. For example a Class table can have student id column but not the Student record itself. The application can however have a Class object with an embedded Student object where Class object has all fields from Class table and the embedded Student object has all fields for the corresponding student in the Student table.

The entity framework is about how you represent the relation in your object programming model and call functions, utilities on them. This is not exactly the same as database design.

Database design is performed by DB architects and they will be more concerned about the performance(read/write), backups, table access privileges.

The application (middle layer) should be able to retrieve info and organize the way it meets the business logic.

During data modeling, the business requirements are taken into consideration, however the design will not be solely based on the application programming approach taken, because if it did then it won't be effective when the software (application) is replaced with another vendor. 
In other words, design data model that makes sense to logically structure data pieces and maintenance and security, and then write your application as per this design and business use case.

Revisiting the original question, it's about database design and I personally think that isolating tables by their context makes them more secure, manageable, scalable.
The object-relational model is about application programming logic and it doesn't address the issues I illustrated in my previous post.

**Comment by SridharSubramaniam** [comment]
Nope. I am referring to core database modelling. Party Model is most wide
used data model (in database).

**Comment by dukesnuz** [comment]
@ysrinu @SridharSubramaniam  Many thanks for all the input. @ysrinu if I am understanding correct... It would be best to separate  each table thus I should continue the data structure per my diagram?  Maybe change address to location.

 As of now all the address tables are identical.. this may change. As for shippers, this table will include customers and receivers(aka ship to address) . The reason I am combining these into one table is that they are interchangeable.  A shipper can be a customer one day and a receiver another day or just a customer.  I have a column called credit which will distinguish whether a shipper or receiver is also a customer.

**Comment by ysrinu** [comment]
@dukesnuz , Sridhar has a good suggestion, please check if that fits your business model.


If I understand your requirement:
A merchant is a supplier such as Amazon, Apple, Target, Walmart
A customer is on the receiving side of the goods.
A customer may also ship to another customer

Therefore, a shipper can mean both a merchant and in some cases a customer.
A merchant business info will have different structure than a customer personal info (ignoring address here)
For example, a business will have LLC name, Tax ID, ...
Customer will have first name, last name

A merchant address will be different than a customer address
A merchant address can have store/location id, branch name, ..

Can a merchant have multiple address? Possibly, for example Walmart can have multiple shipping centers (we are not considering the walmart stores)
A customer can have multiple address, usually with one as the default.

Let's say a customer has multiple addresses, when we have to update an existing address, do you edit the existing address record (i.e. keep the address key the same) or flag this record as 'delete' and add a new address record with new primary key?
On address edit if you edit an existing address record, then for past transactions that used the same key will be 'incorrect' because when you query to get transaction details of this old transaction, it will return address detail as of now and not as of that transaction date. One workaround is to keep the transaction details in a separate table that has full address info as of that time.

table - columns

Merchant:
ID (PK)
LLC_name
TaxId

MerchantAddress:
ID (PK)
Address Line1
Zip
City
State
Store Id

Merchant_MerchantAddress: one-to-many

Customer:
ID (PK)
First Name
Last Name
Default_Address_ID (FK to CustomerAddress table)

CustomerAddress:
ID (PK)
Address Line1
Zip
City
State
Type (Residential, business)


Customer_CustomerAddress: one-to-many

Shipper_Transaction
ID
Shipper_Type (M-Merchant or C-Customer)
Shipper_ID (if shipper_type is 'M' then Merchant.ID else Customer.ID)
address_ID (MerchantAddress.ID or CustomerAddress.ID)

You can also split this into two tables Merchant_Transaction and Customer_Transaction, if makes it easy for your business logic to avoid checking if merchant or not at record level.


To make it simpler, you can merge MerchantAddress and CustomerAddress into one table (some columns may not be applicable to one type or another)
Even if these two address tables are merged into one, I will still prefer to keep them seperate for Shipper, Carrier and Dispatcher


Take sometime to analyze the nature of business, transactions, maintenance and security, and how easy it would be to make future changes to the data model or application logic to adapt to the changing business requirements.


Start with a simple model (not too simple that it will break quickly for new business conditions) and refine as needed.

**Comment by dukesnuz** [comment]
@ysrinu  Many Thanks .... I like your point about editing an address. Most likely I will use soft delete and add a new address. 

All transactions(shipments) areB2B. 
I am going to have a transaction table for each shipment, as you described along with a few more columns.
